File: library/src/main/java/com/devbrackets/android/androidmarkup/text/style/ListSpan.java
package com.devbrackets.android.androidmarkup.text.style;

import android.annotation.SuppressLint;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.os.Build;
import android.text.Layout;
import android.text.style.LeadingMarginSpan;

public class ListSpan implements LeadingMarginSpan {
    public static final int DEFAULT_GAP_WIDTH = 40; //px
    public static final int DEFAULT_BULLET_RADIUS = 4; //px

    public enum Type {
        BULLET,
        NUMERICAL
    }

    private static Path bulletPath = null;

    protected final int gapWidth;
    protected final int bulletRadius;
    protected final Type type;

    //Used for the Numerical list
    private int number = 0;
    private int lastBaseline = -1;

    public ListSpan() {
        this(Type.BULLET, DEFAULT_GAP_WIDTH, DEFAULT_BULLET_RADIUS);
    }

    public ListSpan(int gapWidth) {
        this(Type.BULLET, gapWidth, DEFAULT_BULLET_RADIUS);
    }

    public ListSpan(Type type) {
        this(type, DEFAULT_GAP_WIDTH, DEFAULT_BULLET_RADIUS);
    }

    public ListSpan(Type type, int gapWidth) {
        this(type, gapWidth, DEFAULT_BULLET_RADIUS);
    }

    public ListSpan(Type type, int gapWidth, int bulletRadius) {
        this.gapWidth = gapWidth;
        this.bulletRadius = bulletRadius;
        this.type = type;
    }

    @Override
    public int getLeadingMargin(boolean first) {
        return 2 * bulletRadius + gapWidth;
    }

    @Override
    public void drawLeadingMargin(Canvas canvas, Paint paint, int marginPosition, int direction, int top, int baseline, int bottom, CharSequence text, int start, int end, boolean first, Layout layout) {
        //Cache the style so our changes don't affect others
        Paint.Style style = paint.getStyle();

        if (type == Type.BULLET) {
            drawBulletMargin(canvas, paint, marginPosition, direction, top, baseline, bottom);
        } else {
            drawNumericalMargin(canvas, paint, marginPosition, direction, top, baseline, bottom);
        }

        paint.setStyle(style);
    }

    public Type getType() {
        return type;
    }

    @SuppressLint("NewApi")
    protected void drawBulletMargin(Canvas canvas, Paint paint, int marginPosition, int direction, int top, int baseline, int bottom) {
        paint.setStyle(Paint.Style.FILL);
        int verticalCenter = (top + bottom) / 2;

        //If we don't support hardware acceleration just draw the circle
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB || !canvas.isHardwareAccelerated()) {
            canvas.drawCircle(marginPosition + direction * bulletRadius, verticalCenter, bulletRadius, paint);
            return;
        }

        //Since we support hardware acceleration make sure we have a vector to draw
        if (bulletPath == null) {
            bulletPath = new Path();
            // Bullet is slightly better to avoid aliasing artifacts on mdpi devices.
            bulletPath.addCircle(0.0f, 0.0f, bulletRadius, Path.Direction.CW);
        }

        canvas.save();
        canvas.translate(marginPosition + direction * bulletRadius, verticalCenter);
        canvas.drawPath(bulletPath, paint);
        canvas.restore();
    }

    @SuppressLint("NewApi")
    protected void drawNumericalMargin(Canvas canvas, Paint paint, int marginPosition, int direction, int top, int baseline, int bottom) {
        if (baseline > lastBaseline) {
            lastBaseline = baseline;
            number++;
        } else if (baseline < lastBaseline) {
            lastBaseline = baseline;
            number = 1;
        }

        canvas.drawText(number + ".", marginPosition + direction * bulletRadius, baseline, paint);
    }
}

File: library/src/main/kotlin/com/devbrackets/android/androidmarkup/text/style/ListSpan.kt
package com.devbrackets.android.androidmarkup.text.style

import android.annotation.SuppressLint
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.os.Build
import android.text.Layout
import android.text.style.LeadingMarginSpan

class ListSpan @JvmOverloads constructor(val type: ListSpan.Type = ListSpan.Type.BULLET, protected val gapWidth: Int = ListSpan.DEFAULT_GAP_WIDTH, protected val bulletRadius: Int = ListSpan.DEFAULT_BULLET_RADIUS) : LeadingMarginSpan {

    enum class Type {
        BULLET,
        NUMERICAL
    }

    //Used for the Numerical list
    private var number = 0
    private var lastBaseline = -1

    constructor(gapWidth: Int) : this(Type.BULLET, gapWidth, DEFAULT_BULLET_RADIUS) { }

    override fun getLeadingMargin(first: Boolean): Int {
        return 2 * bulletRadius + gapWidth
    }

    override fun drawLeadingMargin(canvas: Canvas, paint: Paint, marginPosition: Int, direction: Int, top: Int, baseline: Int, bottom: Int, text: CharSequence, start: Int, end: Int, first: Boolean, layout: Layout) {
        //Cache the style so our changes don't affect others
        val style = paint.style

        if (type == Type.BULLET) {
            drawBulletMargin(canvas, paint, marginPosition, direction, top, baseline, bottom)
        } else {
            drawNumericalMargin(canvas, paint, marginPosition, direction, top, baseline, bottom)
        }

        paint.style = style
    }

    @SuppressLint("NewApi")
    protected fun drawBulletMargin(canvas: Canvas, paint: Paint, marginPosition: Int, direction: Int, top: Int, baseline: Int, bottom: Int) {
        paint.style = Paint.Style.FILL
        val verticalCenter = (top + bottom) / 2

        //If we don't support hardware acceleration just draw the circle
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB || !canvas.isHardwareAccelerated) {
            canvas.drawCircle((marginPosition + direction * bulletRadius).toFloat(), verticalCenter.toFloat(), bulletRadius.toFloat(), paint)
            return
        }

        //Since we support hardware acceleration make sure we have a vector to draw
        if (bulletPath == null) {
            bulletPath = Path()
            // Bullet is slightly better to avoid aliasing artifacts on mdpi devices.
            bulletPath!!.addCircle(0.0f, 0.0f, bulletRadius.toFloat(), Path.Direction.CW)
        }

        canvas.save()
        canvas.translate((marginPosition + direction * bulletRadius).toFloat(), verticalCenter.toFloat())
        canvas.drawPath(bulletPath, paint)
        canvas.restore()
    }

    @SuppressLint("NewApi")
    protected fun drawNumericalMargin(canvas: Canvas, paint: Paint, marginPosition: Int, direction: Int, top: Int, baseline: Int, bottom: Int) {
        if (baseline > lastBaseline) {
            lastBaseline = baseline
            number++
        } else if (baseline < lastBaseline) {
            lastBaseline = baseline
            number = 1
        }

        canvas.drawText("$number.", (marginPosition + direction * bulletRadius).toFloat(), baseline.toFloat(), paint)
    }

    companion object {
        val DEFAULT_GAP_WIDTH = 40 //px
        val DEFAULT_BULLET_RADIUS = 4 //px

        private var bulletPath: Path? = null
    }
}